### PR TITLE
ci(aio): deploy commits on the stable branch to production

### LINKED ISSUE #16908: aio: setup production deployment scripts

### PR DESCRIPTION
**Pull request by gkalpak**
The current stable branch is determined based on the current version mapped to the npm `latest` tag (by replacing the patch version number with 'x' - e.g. `1.2.3` --> `1.2.x`).
PRs against the stable branch will be deployed to the preview server (as long as the rest of the requirements are met). Commits on the stable branch itself will be deployed to production.

Fixes #16908.

### CODE REVIEW

**Comment by gkalpak** [comment]
TODO (before this PR is merged):

- [ ] Use the correct project ID for production. (Currently using `aio-production` as a placeholder.)
- [ ] Use the correct `FIREBASE_TOKEN` for production. (Currently using the staging token as a placeholder.)

File: aio/scripts/deploy-to-firebase.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+(
+  cd "`dirname $0`/.."
+
+  # Build the app
+  yarn build
+
+  # Deploy to Firebase
+  firebase use "$projectId" --token "$firebaseToken"
+  firebase deploy --message "Commit: $TRAVIS_COMMIT" --non-interactive --token "$firebaseToken"
+
```
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: no `sleep 10` any more?

File: scripts/ci/deploy.sh [hunk 2]
```diff
@@ -40,24 +40,33 @@ case ${CI_MODE} in
...
+      (
+        cd ${TRAVIS_BUILD_DIR}/aio
+
+        if [[ $TRAVIS_BRANCH == $STABLE_BRANCH ]]; then
+          # This is upstream <stable-branch>: Deploy to production.
+          travisFoldStart "deploy.aio.production"
+            yarn deploy-production
+          travisFoldEnd "deploy.aio.production"
+        else
+          # This is upstream master: Deploy to staging.
```
Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: Do we need a check for master here?
What if someone builds a branch that is neither stable nor master? Won't it still get pushed to staging?

Comment (petebacondarwin) [inline scripts/ci/deploy.sh]: Oh, no I see that this is covered above.

**Review by petebacondarwin — APPROVED** [review]
LGTM

File: aio/scripts/deploy-to-firebase.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+set +x -eu -o pipefail
+
+
+readonly deployEnv=$1
+
+case $deployEnv in
+  staging)
+    readonly projectId=aio-staging
+    readonly deployedUrl=https://$projectId.firebaseapp.com/
+    readonly firebaseToken=$FIREBASE_TOKEN
```
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: Is it the same firebase token for both deployments? What if it is not?

Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: It's not (see https://github.com/angular/angular/pull/16953#issuecomment-303335448).

File: aio/scripts/deploy-to-firebase.sh (hunk 1 above)
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: It didn't help :smiley:
I still need to debug why lighthouse doesn't work for google.com/firebaseapp.com from Travis.

File: aio/scripts/deploy-to-firebase.sh (hunk 3 above)
Comment (gkalpak) [inline aio/scripts/deploy-to-firebase.sh]: It turns out it is the same (set in #17005).

**Comment by gkalpak** [comment]
The `FIREBASE_TOKEN` has been update in #17005. So, this should be good to go.

**Comment by petebacondarwin** [comment]
Just needs `root` approval.

File: scripts/ci/env.sh [hunk 4]
```diff
@@ -59,6 +59,10 @@ if [[ ${TRAVIS:-} ]]; then
       setEnvVar KARMA_JS_BROWSERS `node -e "console.log(require('/home/travis/build/angular/angular/browser-providers.conf').browserstackAliases.CI_OPTIONAL.join(','))"`
       ;;
     aio)
+      # Determine the current stable branch.
+      readonly latestRe="^\s*latest:\s*([0-9]+\.[0-9]+)\.[0-9]+.*$"
+      setEnvVar STABLE_BRANCH `npm dist-tag ls @angular/core | grep -P $latestRe | tail -1 | sed -r "s/$latestRe/\1.x/"`
```
Comment (IgorMinar) [inline scripts/ci/env.sh]: can you rewrite this into: `npm info @angular/core dist-tags.next` - you won't need any regular expression matching or sed for that.

**Review by IgorMinar — CHANGES_REQUESTED** [review]
the rest looks good. thanks!

File: scripts/ci/env.sh (hunk 4 above)
Comment (gkalpak) [inline scripts/ci/env.sh]: Here we are trying to determine the branch name by replacing the patch version number with `x` (e.g. `4.1.3` --> `4.1.x`). I'm afraid we need `sed` (or do you have other ideas)?

Comment (gkalpak) [inline scripts/ci/env.sh]: Done.

**Review by IgorMinar — APPROVED** [review]

File: aio/scripts/deploy-to-firebase.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    echo "Unknown deployment environment ('$deployEnv'). Expected 'staging' or 'production'."
+    exit 1
+    ;;
+esac
+
+(
+  cd "`dirname $0`/.."
+
+  # Build the app
+  yarn build
```
Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: we need to be able to pass in the environment to this task

Comment (petebacondarwin) [inline aio/scripts/deploy-to-firebase.sh]: How about we break up the `yarn build` task into `prebuild`, `build` and `postbuild`:

```
"prebuild": "yarn check-env && yarn setup",
"build": ng build -prod -sm -vc=false",
"postbuild": "yarn sw-manifest && yarn sw-copy"
```

Then we could do something like:

```
yarn build -- --env=stage
```

Comment (IgorMinar) [inline aio/scripts/deploy-to-firebase.sh]: yeah. I think this is right, but I don't think we actually need `environment.staging.ts`, we can just use the dev `environment.ts` for aio-staging builds.

Comment (IgorMinar) [inline aio/scripts/deploy-to-firebase.sh]: let's do this in a separate PR though. I marked this one for merging
